hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/CallIdSequence.java

```diff
@@ -0,0 +1,249 @@
+
+        private final int maxConcurrentInvocations;
+
+        public CallIdSequenceWithBackpressure(int maxConcurrentInvocations) {
+            if (maxConcurrentInvocations < 1) {
+                throw new IllegalArgumentException("maxConcurrentInvocations can't be smaller than 1");
+            }
+            this.maxConcurrentInvocations = maxConcurrentInvocations;
+        }
+
+        @Override
+        public long getLastCallId() {
+            return longs.get(INDEX_HEAD);
+        }
+
+        long getTail() {
+            return longs.get(INDEX_TAIL);
+        }
+
+        @Override
+        public int getMaxConcurrentInvocations() {
+            return maxConcurrentInvocations;
+        }
+
+        @Override
+        public long next(Invocation invocation) {
+            assert invocation.op.getCallId() == 0 : "callId should be null:" + invocation;
+
+            if (!invocation.op.isUrgent()) {
+                if (!hasSpace()) {
+                    waitForSpace(invocation);
+                }
+            }
+
+            // we need to generate a sequence, no matter if the invocation is going to skip registration.
+            // because this information is used to determine the number of in flight invocations
+            long callId = next();
+
+            if (skipRegistration(invocation)) {
+                return CALL_ID_LOCAL_SKIPPED;
+            } else {
+                return callId;
+            }
+        }
+
+        private long next() {
+            return longs.incrementAndGet(INDEX_HEAD);
+        }
+
+        private boolean hasSpace() {
+            return longs.get(INDEX_HEAD) - longs.get(INDEX_TAIL) < maxConcurrentInvocations;
+        }
+
+        private void waitForSpace(Invocation invocation) {
+            long remainingTimeoutMs = invocation.op.getCallTimeout();
+            boolean restoreInterrupt = false;
+            try {
+                long delayMs = 1;
+                for (; ; ) {
+                    long startMs = System.currentTimeMillis();
+                    try {
+                        Thread.sleep(delayMs);
+                    } catch (InterruptedException e) {
+                        restoreInterrupt = true;
+                    }
+                    long durationMs = System.currentTimeMillis() - startMs;
+                    remainingTimeoutMs -= durationMs;
+
+                    if (hasSpace()) {
+                        return;
+                    }
+
+                    if (remainingTimeoutMs <= 0) {
+                        remainingTimeoutMs = 0;
+                        throw new HazelcastOverloadException("Failed to get a callId for invocation: " + invocation);
+                    }
+
+                    delayMs = nextDelay(remainingTimeoutMs, delayMs);
+                    remainingTimeoutMs -= delayMs;
+                }
+            } finally {
+                setCallTimeout(invocation.op, remainingTimeoutMs);
```

Comment:
I'm not sure if setting remaining time as call-timeout. Because call-timeout is used as timeout of each remote call, it means for each retry not the whole invocation. 

For example normally, using 1min call-timeout, a caller can wait infinitely as long as a retry-response is received in every 1min. But when no response is received in call-timeout (+ some extra time), invocation will throw operation-timeout exception. 

If you set remaining-time as call-timeout, on every retry call-timeout will decrease until it reaches zero. I don't know if this is your intention...